src/lib/server/endpoints/hfinference/endpointHfInference.ts

```diff
@@ -0,0 +1,294 @@
+import { z } from "zod";
+import type { Endpoint, TextGenerationStreamOutputWithToolsAndWebSources } from "../endpoints";
+import { createImageProcessorOptionsValidator, makeImageProcessor } from "../images";
+import { INFERENCE_PROVIDERS, InferenceClient } from "@huggingface/inference";
+import { config } from "$lib/server/config";
+import type { Tool, ToolCall } from "$lib/types/Tool";
+import type { ChatCompletionStreamOutput } from "@huggingface/tasks";
+import type { FunctionDefinition } from "openai/resources/index.mjs";
+import type { ChatCompletionTool, FunctionParameters } from "openai/resources/index.mjs";
+import { logger } from "$lib/server/logger";
+import type { Message, MessageFile } from "$lib/types/Message";
+
+type DeltaToolCall = NonNullable<
+	ChatCompletionStreamOutput["choices"][number]["delta"]["tool_calls"]
+>[number];
+
+function createChatCompletionToolsArray(tools: Tool[] | undefined): ChatCompletionTool[] {
+	const toolChoices = [] as ChatCompletionTool[];
+	if (tools === undefined) {
+		return toolChoices;
+	}
+
+	for (const t of tools) {
+		const requiredProperties = [] as string[];
+
+		const properties = {} as Record<string, unknown>;
+		for (const idx in t.inputs) {
+			const parameterDefinition = t.inputs[idx];
+
+			const parameter = {} as Record<string, unknown>;
+			switch (parameterDefinition.type) {
+				case "str":
+					parameter.type = "string";
+					break;
+				case "float":
+				case "int":
+					parameter.type = "number";
+					break;
+				case "bool":
+					parameter.type = "boolean";
+					break;
+				case "file":
+					throw new Error("File type's currently not supported");
+				default:
+					throw new Error(`Unknown tool IO type: ${t}`);
+			}
+
+			if ("description" in parameterDefinition) {
+				parameter.description = parameterDefinition.description;
+			}
+
+			if (parameterDefinition.paramType == "required") {
+				requiredProperties.push(t.inputs[idx].name);
+			}
+
+			properties[t.inputs[idx].name] = parameter;
+		}
+
+		const functionParameters: FunctionParameters = {
+			type: "object",
+			...(requiredProperties.length > 0 ? { required: requiredProperties } : {}),
+			properties,
+		};
+
+		const functionDefinition: FunctionDefinition = {
+			name: t.name,
+			description: t.description,
+			parameters: functionParameters,
+		};
+
+		const toolDefinition: ChatCompletionTool = {
+			type: "function",
+			function: functionDefinition,
+		};
+
+		toolChoices.push(toolDefinition);
+	}
+
+	return toolChoices;
+}
+
+export const endpointHfInferenceParametersSchema = z.object({
+	type: z.literal("hfinference"),
```

Comment:
i would name it either `inference-providers` (the name of the product) or `InferenceClient` (the name of the library class)

Reply:
otherwise it's confusing with `provider="hf-inference"`